File: android/src/main/java/com/netease/nertcflutter/NERtcChannelManager.java
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

package com.netease.nertcflutter;

import android.text.TextUtils;
import android.util.Log;
import androidx.annotation.NonNull;
import com.netease.lava.nertc.sdk.NERtcEx;
import com.netease.lava.nertc.sdk.channel.NERtcChannel;
import com.netease.lava.nertc.sdk.channel.NERtcChannelCallback;
import com.netease.lava.nertc.sdk.stats.NERtcStatsObserver;
import io.flutter.plugin.common.BinaryMessenger;
import java.util.HashMap;
import java.util.Map;

public class NERtcChannelManager {

  private static final String TAG = "NERtcChannelManager";
  private static volatile NERtcChannelManager instance;

  private final Map<String, NERtcChannel> channelMaps = new HashMap<>();
  private final Map<String, NERtcChannelCallback> callbackMaps = new HashMap<>();
  private final Map<String, NERtcStatsObserver> statsObservers = new HashMap<>();
  private BinaryMessenger messenger;

  public static NERtcChannelManager getInstance() {
    if (instance == null) {
      synchronized (NERtcChannelManager.class) {
        if (instance == null) {
          instance = new NERtcChannelManager();
        }
      }
    }
    return instance;
  }

  public void setBinaryMessager(BinaryMessenger messenger) {
    this.messenger = messenger;
  }

  public long createChannel(@NonNull String channelName) {

    if (TextUtils.isEmpty(channelName)) {
      Log.e(TAG, "Channel name is empty");
      return -1L;
    }

    if (channelMaps.containsKey(channelName)) {
      Log.d(TAG, "Channel already exists: " + channelName);
      return 0L;
    }

    NERtcChannel channel = NERtcEx.getInstance().createChannel(channelName);
    if (channel == null) {
      Log.e(TAG, "Failed to create channel: " + channelName);
      return -1L;
    }

    // 保存到映射表
    channelMaps.put(channelName, channel);
    NERtcSubCallbackImpl callback = new NERtcSubCallbackImpl(channelName, this.messenger);
    channel.setChannelCallback(callback);
    callbackMaps.put(channelName, callback);
    return 0L;
  }

  public NERtcChannel getChannel(String channelName) {
    if (TextUtils.isEmpty(channelName)) {
      return null;
    }
    return channelMaps.get(channelName);
  }

  public NERtcChannelCallback getCallback(String channelName) {
    if (TextUtils.isEmpty(channelName)) {
      return null;
    }
    return callbackMaps.get(channelName);
  }

  public void release(String channelName) {
    if (TextUtils.isEmpty(channelName)) {
      return;
    }
    NERtcChannel channel = channelMaps.remove(channelName);
    if (channel != null) {
      channel.setChannelCallback(null);
      channel.release();
    }
    callbackMaps.remove(channelName);
    statsObservers.remove(channelName);
  }

  public void releaseAll() {

    //遍历所有 channel 并销毁
    for (NERtcChannel channel : channelMaps.values()) {
      if (channel != null) {
        channel.setChannelCallback(null);
        channel.release();
      }
    }

    // 清空所有映射表
    channelMaps.clear();
    callbackMaps.clear();
    statsObservers.clear();
  }

  public NERtcStatsObserver createStatsObserver(String channelName) {
    if (TextUtils.isEmpty(channelName)) {
      Log.e(TAG, "Channel name is empty");
      return null;
    }

    if (statsObservers.containsKey(channelName)) {
      return statsObservers.get(channelName);
    }

    NERtcStatsObserver observer = new NERtcStatsObserverImpl(this.messenger, channelName);
    statsObservers.put(channelName, observer);
    return observer;
  }

  public NERtcStatsObserver getStatsObserver(String channelName) {
    if (TextUtils.isEmpty(channelName)) {
      return null;
    }
    return statsObservers.get(channelName);
  }
}

File: ios/Classes/NERtcChannelManager.m
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

#import "NERtcChannelManager.h"
#import <NERtcSDK/NERtcSDK.h>
#import "NERtcStatsObserverImpl.h"
#import "NERtcSubCallbackImpl.h"

@interface NERtcChannelManager ()
@property(nonatomic, strong) NSMutableDictionary<NSString *, NERtcChannel *> *channelMaps;
@property(nonatomic, strong)
    NSMutableDictionary<NSString *, id<NERtcChannelDelegate>> *callbackMaps;
@property(nonatomic, strong)
    NSMutableDictionary<NSString *, id<NERtcChannelMediaStatsObserver>> *statsObservers;
@property(nonatomic, weak) NSObject<FlutterBinaryMessenger> *messenger;
@end

@implementation NERtcChannelManager

static NERtcChannelManager *_instance = nil;
static dispatch_once_t onceToken;

+ (instancetype)sharedInstance {
  dispatch_once(&onceToken, ^{
    _instance = [[NERtcChannelManager alloc] init];
  });
  return _instance;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    _channelMaps = [[NSMutableDictionary alloc] init];
    _callbackMaps = [[NSMutableDictionary alloc] init];
    _statsObservers = [[NSMutableDictionary alloc] init];
  }
  return self;
}

- (void)setBinaryMessenger:(NSObject<FlutterBinaryMessenger> *)messenger {
  self.messenger = messenger;
}

- (NSInteger)createChannel:(NSString *)channelName {
  if (!channelName || channelName.length == 0) {
    NSLog(@"[NERtcChannelManager] Channel name is empty");
    return -1;
  }

  if ([self.channelMaps objectForKey:channelName]) {
    NSLog(@"[NERtcChannelManager] Channel already exists: %@", channelName);
    return 0;
  }

  NERtcChannel *channel = [[NERtcEngine sharedEngine] createChannel:channelName];
  if (!channel) {
    NSLog(@"[NERtcChannelManager] Failed to create channel: %@", channelName);
    return -1;
  }

  // 保存到映射表
  [self.channelMaps setObject:channel forKey:channelName];

  // 创建并设置回调
  NERtcSubCallbackImpl *callback =
      [[NERtcSubCallbackImpl alloc] initWithChannelName:channelName binaryMessenger:self.messenger];
  [channel setChannelDelegate:callback];
  [self.callbackMaps setObject:callback forKey:channelName];

  return 0;
}

- (nullable NERtcChannel *)getChannel:(NSString *)channelName {
  if (!channelName || channelName.length == 0) {
    return nil;
  }
  return [self.channelMaps objectForKey:channelName];
}

- (nullable id<NERtcChannelDelegate>)getCallback:(NSString *)channelName {
  if (!channelName || channelName.length == 0) {
    return nil;
  }
  return [self.callbackMaps objectForKey:channelName];
}

- (void)releaseChannel:(NSString *)channelName {
  if (!channelName || channelName.length == 0) {
    return;
  }

  NERtcChannel *channel = [self.channelMaps objectForKey:channelName];
  if (channel) {
    [channel setChannelDelegate:nil];
    [channel destroy];
    [self.channelMaps removeObjectForKey:channelName];
  }

  [self.callbackMaps removeObjectForKey:channelName];
  [self.statsObservers removeObjectForKey:channelName];
}

- (void)releaseAll {
  // 清空所有 channel 并销毁
  for (NSString *channelName in [self.channelMaps allKeys]) {
    NERtcChannel *channel = [self.channelMaps objectForKey:channelName];
    if (channel) {
      [channel setChannelDelegate:nil];
      [channel destroy];
    }
  }

  // 清空所有映射表
  [self.channelMaps removeAllObjects];
  [self.callbackMaps removeAllObjects];
  [self.statsObservers removeAllObjects];
}

- (nullable id<NERtcChannelMediaStatsObserver>)createStatsObserver:(NSString *)channelName {
  if (!channelName || channelName.length == 0) {
    NSLog(@"[NERtcChannelManager] Channel name is empty");
    return nil;
  }

  id<NERtcChannelMediaStatsObserver> existingObserver =
      [self.statsObservers objectForKey:channelName];
  if (existingObserver) {
    return existingObserver;
  }

  NERtcStatsObserverImpl *observer =
      [[NERtcStatsObserverImpl alloc] initWithBinaryMessenger:self.messenger
                                                   channelTag:channelName];
  [self.statsObservers setObject:observer forKey:channelName];
  return observer;
}

- (nullable id<NERtcChannelMediaStatsObserver>)getStatsObserver:(NSString *)channelName {
  if (!channelName || channelName.length == 0) {
    return nil;
  }
  return [self.statsObservers objectForKey:channelName];
}

@end
